### PR TITLE
Quad to triangle optimization

### PR DESCRIPTION
**Pull request by BaronKiko**
Optimizes quad rendering by converting to triangles when uploading (reduces gpu load by 2/3rds in Shantae)
Also fixes minor bug regarding IndexFirst not being converted correctly with quad and quad strips

### CODE REVIEW

**Comment by BaronKiko** [comment]
Should mention quadstrips are untested as xenoblade is the only game I know of that uses them and I don't have it nor do I think it would boot enough to test.
If there are issues they are probably in the preexisting quad helper class.

**Comment by BaronKiko** [comment]
A shiny blank line has been added for code formatting reasons.

File: Ryujinx.Graphics/Graphics3d/NvGpuEngine3d.cs [hunk 1]
```diff
@@ -796,19 +796,37 @@ private void UploadVertexArrays(NvGpuVmm Vmm, GalPipelineState State)
...
                 long VbSize = (VbEndPos - VbPosition) + 1;
+                int ModifiedVbSize = (int)VbSize;
 
-                bool VboCached = Gpu.Renderer.Rasterizer.IsVboCached(VboKey, VbSize);
+
+                // If quads convert size to triangle length
+                if (Stride == 0)
+                {
+                    if (PrimType == GalPrimitiveType.Quads)
+                        ModifiedVbSize = QuadHelper.ConvertIbSizeQuadsToTris(ModifiedVbSize);
```
Comment (gdkchan) [inline Ryujinx.Graphics/Graphics3d/NvGpuEngine3d.cs]: Please add braces `{` `}` on the if/else.

File: Ryujinx.Graphics/Graphics3d/NvGpuEngine3d.cs [hunk 2]
```diff
@@ -796,19 +796,37 @@ private void UploadVertexArrays(NvGpuVmm Vmm, GalPipelineState State)
...
                 if (!VboCached || Gpu.ResourceManager.MemoryRegionModified(Vmm, VboKey, VbSize, NvGpuBufferType.Vertex))
                 {
-                    if (Vmm.TryGetHostAddress(VbPosition, VbSize, out IntPtr VbPtr))
+                    if ((PrimType == GalPrimitiveType.Quads | PrimType == GalPrimitiveType.QuadStrip) && Stride != 0)
                     {
-                        Gpu.Renderer.Rasterizer.CreateVbo(VboKey, (int)VbSize, VbPtr);
+                        // Convert quad buffer to triangles
+                        byte[] data = Vmm.ReadBytes(VbPosition, VbSize);
+
+                        if (PrimType == GalPrimitiveType.Quads)
```
Comment (gdkchan) [inline Ryujinx.Graphics/Graphics3d/NvGpuEngine3d.cs]: Ditto.

File: Ryujinx.Graphics/Graphics3d/NvGpuEngine3d.cs [hunk 3]
```diff
@@ -863,22 +881,17 @@ private void DispatchRender(NvGpuVmm Vmm, GalPipelineState State)
...
-                if (PrimType == GalPrimitiveType.Quads ||
-                    PrimType == GalPrimitiveType.QuadStrip)
+                if (PrimType == GalPrimitiveType.Quads || PrimType == GalPrimitiveType.QuadStrip)
                 {
-                    PrimType = GalPrimitiveType.Triangles;
-
                     //Note: We assume that index first points to the first
                     //vertex of a quad, if it points to the middle of a
                     //quad (First % 4 != 0 for Quads) then it will not work properly.
                     if (PrimType == GalPrimitiveType.Quads)
```
Comment (gdkchan) [inline Ryujinx.Graphics/Graphics3d/NvGpuEngine3d.cs]: Ditto.

File: Ryujinx.Graphics/Graphics3d/NvGpuEngine3d.cs [hunk 4]
```diff
@@ -888,6 +901,23 @@ private void DispatchRender(NvGpuVmm Vmm, GalPipelineState State)
...
 
+                //Quad primitive types were deprecated on OpenGL 3.x,
+                //they are converted to a triangles index buffer on IB creation,
+                //so we should use the triangles type here too.
+                if (PrimType == GalPrimitiveType.Quads || PrimType == GalPrimitiveType.QuadStrip)
+                {
+                    //Note: We assume that index first points to the first
+                    //vertex of a quad, if it points to the middle of a
+                    //quad (First % 4 != 0 for Quads) then it will not work properly.
+                    if (PrimType == GalPrimitiveType.Quads)
```
Comment (gdkchan) [inline Ryujinx.Graphics/Graphics3d/NvGpuEngine3d.cs]: Ditto.

File: Ryujinx.Graphics/Graphics3d/NvGpuEngine3d.cs [hunk 5]
```diff
@@ -796,19 +796,37 @@ private void UploadVertexArrays(NvGpuVmm Vmm, GalPipelineState State)
...
+                        // Convert quad buffer to triangles
+                        byte[] data = Vmm.ReadBytes(VbPosition, VbSize);
+
+                        if (PrimType == GalPrimitiveType.Quads)
+                            data = QuadHelper.ConvertIbQuadsToTris(data, Stride, (int)(VbSize / Stride));
+                        else
+                            data = QuadHelper.ConvertIbQuadStripToTris(data, Stride, (int)(VbSize / Stride));
+                        Gpu.Renderer.Rasterizer.CreateVbo(VboKey, data);
                     }
+                    else if (Vmm.TryGetHostAddress(VbPosition, VbSize, out IntPtr VbPtr))
```
Comment (gdkchan) [inline Ryujinx.Graphics/Graphics3d/NvGpuEngine3d.cs]: Ditto.

File: Ryujinx.Graphics/Graphics3d/NvGpuEngine3d.cs [hunk 6]
```diff
@@ -796,19 +796,37 @@ private void UploadVertexArrays(NvGpuVmm Vmm, GalPipelineState State)
...
+                        Gpu.Renderer.Rasterizer.CreateVbo(VboKey, (int)VbSize, VbPtr);
                     else
```
Comment (gdkchan) [inline Ryujinx.Graphics/Graphics3d/NvGpuEngine3d.cs]: Ditto.

**Comment by BaronKiko** [comment]
I think dropping the Ib should be clear enough given it's a part of QuadHelper. I'm stopping for the day so will sort it out tomorrow.

**Comment by BaronKiko** [comment]
A band of wild brackets has appeared. Also some QuadHelper refactoring.

**Comment by EmulationChannel** [comment]
sharedscreenshot

**Comment by BaronKiko** [comment]
I can't even get in game nor do I know if this is worse, better or the same. Can you give a little context?

**Comment by gdkchan** [comment]
I think the screenshot is out of context, the change shouldn't affect SMO at all.

**Comment by BaronKiko** [comment]
Yeah that was my thinking. Would still like to know why @EmulationChannel decided to post it here.

**Comment by EmulationChannel** [comment]
Game FPS 0.00 SMO

**Comment by Thog** [comment]
@EmulationChannel This is a known issue that is not related to this PR issue. (This also happens on master)

**Comment by gdkchan** [comment]
It was like this for a while on nvidia, again I don't see how its related to this change.

**Comment by BaronKiko** [comment]
Maybe @EmulationChannel was testing performance because of "reduces gpu load by 2/3rds in Shantae" in the original post? I should make it clear that gain is only on games that use quads which is very few. Anyhow I too think it's unrelated.

**Pull request merged**